### ISSUE TITLE
Home feed exhausts RAM

### ISSUE THREAD

**Issue by sevenbitbyte** [mastodon/mastodon#21677]
## Repro Steps

1. Create an account on a server using the default webapp.
2. Follow some active accounts
3. Leave the tab open and idle for 8hr+
4. Check RAM usage

## Outcome

The webapp appears to store in RAM all posts that showed up during this time period. In my case I had multiple tabs open so the app tabs had claimed upwards of 6GB during this time. The application does not appear to be RAM aware and had fully exhausted my computers swap space and most of the system RAM. Had it continued Chrome or my OS likely would have called the app. On some OSes this will result in a complete system deadlock once working RAM is exhausted.

## Suggestions

1. Consider sharing home feed data between tabs so tabs do not duplicate so much content
2. Is the DOM attempting to render off screen content? If so please utilize a true "Infinite Scroll" implementation which does not populate out of view DOM content.
3. Some Mastodon users have suggested enabling "slow mode", RAM exhaustion shouldn't be the default setting. Consider making "slow mode" the default if this is how infinite scroll is currently supported.
4. Store posts content in IndexDB rather than RAM. (I'm still learning to determine where post content is stored. It appears to only be RAM as far as I can see so far)

### Expected behaviour

RAM not exhausted

### Actual behaviour

RAM exhausted

### Specifications

Google Chrome 107.0.5304.110
Mastodon v4.0.2
Host: fosstodon.org

**Comment by afontenot** [comment]
Yep this is an annoying issue. It's related to similar problems on the public timelines, which behave exactly the same way despite being even faster than the home timeline. https://github.com/mastodon/mastodon/issues/6438 https://github.com/mastodon/mastodon/issues/21486

Slow mode actually doesn't solve this problem although it may lessen it. Slow mode still downloads new posts into memory, it just doesn't add them to the page.

I think the reasonable approach here would be to stop downloading new posts after the user has been inactive for 10 minutes, or 250 posts have been downloaded but not viewed or displayed (or something like that).

**Comment by ke352802081770314** [comment]
Indeed all posts are stored in the HTML body of the page/tab displaying /home (or some other timeline). There seems to be no clean-up at all.
If I keep the /home tab open for a few weeks and read the posts in forward time order, i.e. by scrolling up while the top gets filled in with new posts, the "tail" of already read posts just keeps growing and is never cut off, until I close or navigate away from /home. I can see that by the fact that the position of the thumb in the scrollbar is quite high up. What's more, when I do click on a post and thereby replace /home by it, intentionally or not, and I go "back", I am at mercy of the browser's cache whether I will see the timeline I just left, or whether the browser decided to throw away the cached content and loads it from the server - which may not be able to give me the posts I just saw any more due to the 800 posts limit, if my backlog was bigger than that. If it kept the cache, for several minutes the timeline may wobble up and down as it re-renders the posts and recalculates the height of the page all the time.

At one point, the HTML for the post I was viewing started as follows:
`<article aria-posinset="846" aria-setsize="5046" data-id="115949862232224632" tabindex="-1" style="">` 
the Firefox process for the mastodon server was at 12GB size. As you can see I was at post 846 from the top, so I was dependent on the /home tab surviving as I would not be able to reproduce the timeline up to that point from the server. However, after the point I was at, more than 4000 older posts remain in the HTML that I am unlikely to revisit.
In addition to the memory, this also consumes a lot of CPU, as the Javascript has to adjust each aria-posinset and aria-setsize by incrementing every time a new post comes in, also it has to adjust all the relative timestamps - "aging" them - periodically. The CPU load is illustrated when I compose a post. The characters in the entry field appear like one per second, so I can type ahead and then watch as the text slowly emerges. This is hardly usable.

Meanwhile, knowing about these shortcomings, I have started the habit of opening a new /home tab every couple of days, and switch to that when my reading reaches the post present in the new one, then I close the old /home tab to free up ressources. This should not be a task left to the user. At the very least, I wish there was an option to cut off all posts e.g. more than 500 below where I am currently reading, when I want to do so.

As I explained, due to the 800 limit, not adding the new posts to the timeline is not an option that would help me, because by the time I reach the top of the posts already loaded earlier, the subsequent ones may no longer be retrievable from the server.

**Comment by velw** [comment]
This could explain why the remaining space on my hard disk is consumed by the browser if I leave Mastodon open in a window all day. When I close the browser I get the space back. I usually notice because the lack of disk space has stopped something else from working. A little concerning that it was flagged in 2022 and doesn't seem to have been triaged.

**Comment by diondiondion** [comment]
We're definitely aware of this issue & would like to fix it, unfortunately it's quite complex and will require some planning.
